usr.bin/sockstat/sockstat.c

```diff
@@ -1345,65 +1368,104 @@ display_sock(struct sock *s, struct col_widths *cw, char *buf, size_t bufsize)
 	laddr = s->laddr;
 	faddr = s->faddr;
 	first = true;
+	const bool is_text_style = (xo_get_style(NULL) == XO_STYLE_TEXT);
 
 	snprintf(buf, bufsize, "%s%s%s",
 		s->protoname,
 		s->vflag & INP_IPV4 ? "4" : "",
 		s->vflag & INP_IPV6 ? "6" : "");
-	printf(" %-*s", cw->proto, buf);
+	xo_emit(" {:proto/%-*s}", cw->proto, buf);
 	while (laddr != NULL || faddr != NULL) {
 		if (s->family == AF_UNIX) {
 			if ((laddr == NULL) || (faddr == NULL))
-				errx(1, "laddr = %p or faddr = %p is NULL",
+				xo_errx(1, "laddr = %p or faddr = %p is NULL",
 					(void *)laddr, (void *)faddr);
-			if (laddr->address.ss_len > 0)
+			if (laddr->address.ss_len > 0) {
+				xo_open_container("local");
 				formataddr(&laddr->address, buf, bufsize);
-			else if (laddr->address.ss_len == 0 && faddr->conn == 0)
-				strlcpy(buf, "(not connected)", bufsize);
-			else
-				strlcpy(buf, "??", bufsize);
-			printf(" %-*.*s", cw->local_addr, cw->local_addr, buf);
-			if (format_unix_faddr(faddr, buf, bufsize) == 0)
-				strlcpy(buf, "??", bufsize);
-			printf(" %-*.*s", cw->foreign_addr,
-				cw->foreign_addr, buf);
+				if (is_text_style) {
+					xo_emit(" {:/%-*.*s}", cw->local_addr,
+						cw->local_addr, buf);
+				}
+				xo_close_container("local");
+			} else if (laddr->address.ss_len == 0 &&
+				faddr->conn == 0 && is_text_style) {
+				xo_emit(" {:/%-*.*s}", cw->local_addr,
+					cw->local_addr, "(not connected)");
+			} else if (is_text_style) {
+				xo_emit(" {:/%-*.*s}", cw->local_addr,
+					cw->local_addr, "??");
+			}
+			if (faddr->conn != 0 || faddr->firstref != 0) {
+				xo_open_container("foreign");
+				int len = format_unix_faddr(faddr, buf,
+						bufsize);
+				if (len == 0 && is_text_style)
+					xo_emit(" {:/%-*s}",
+						cw->foreign_addr, "??");
+				else if (is_text_style)
+					xo_emit(" {:/%-*.*s}", cw->foreign_addr,
+						cw->foreign_addr, buf);
+				xo_close_container("foreign");
+			} else if (is_text_style)
+				xo_emit(" {:/%-*s}", cw->foreign_addr, "??");
 		} else {
-			if (laddr != NULL)
+			if (laddr != NULL) {
+				xo_open_container("local");
 				formataddr(&laddr->address, buf, bufsize);
-			else
-				strlcpy(buf, "??", bufsize);
-			printf(" %-*.*s", cw->local_addr, cw->local_addr, buf);
-			if (faddr != NULL)
+				if (is_text_style) {
+					xo_emit(" {:/%-*.*s}", cw->local_addr,
+						cw->local_addr, buf);
+				}
+				xo_close_container("local");
+			} else if (is_text_style)
+				xo_emit(" {:/%-*.*s}", cw->local_addr,
+					cw->local_addr, "??");
+			if (faddr != NULL) {
+				xo_open_container("foreign");
```

Comment:
Also here, for consistency's sake, there should be a `xo_open_container("connections")`.